kernel: Remove UniValue from kernel library

sedited opened the pull request:
Besides the build system changes, this is a mostly move-only change for moving the few UniValue-related functions out of kernel files.

UniValue is not required by any of the kernel components and a JSON library should not need to be part of a consensus library.

hebasto commented:
Concept ACK.

stickies-v commented:
Concept ACK

maflcko reviewed (approved):
lgtm. Left two nits, feel free to ignore.


lgtm ACK 03e06e589a24f0bcb7dbd28a0aaa925b18b87b9d 🚆

<details><summary>Show signature</summary>

Signature:

```
untrusted comment: signature from minisign secret key on empty file; verify via: minisign -Vm "${path_to_any_empty_file}" -P RWTRmVTMeKV5noAMqVlsMugDDCyyTSbA3Re5AkUrhvLVln0tSaFWglOw -x "${path_to_this_whole_four_line_signature_blob}"
RUTRmVTMeKV5npGrKx1nqXCw5zeVHdtdYURB/KlyA/LMFgpNCs+SkW9a8N95d+U4AP1RJMi+krxU1A3Yux4bpwZNLvVBKy0wLgM=
trusted comment: lgtm ACK 03e06e589a24f0bcb7dbd28a0aaa925b18b87b9d 🚆
2mxQ9s2E83Cpsc3G1cPpXHudEp5PcMmrgHTI+WVd+CI94t4cNzqP53ApM7yvTxGkPfWXNc8sndlQVaImSA5DAg==
```

</details>

stickies-v commented on src/common/univalue_helpers.cpp:
```diff
@@ -0,0 +1,47 @@
+    return ParseHex(strHex);
+}
+
+int ParseSighashString(const UniValue& sighash)
```
I don't really find this fn to be a `UniValue` helper, pretty much all the logic is sighash specific. While thinking for a better place to sit, I ended up thinking the better solution may be to keep it where it is but just remove the (imo unnecessary) dependency on UV? Offloading the default value and type check to a helper function like `ParseSighashString` is opaque, and even though my approach here is slightly more verbose in the callsite I find the clarity of what's happening to be worth it (different callsites may require different behaviour in case of wrong types or `sighash_str`).

What do you think?

_(I think returning a `std::optional` instead of throwing is probably even better, but orthogonal to this pull)_

<details>
<summary>git diff on master</summary>

```diff
diff --git a/src/core_io.h b/src/core_io.h
index 997f3bfd5b..ea07042b82 100644
--- a/src/core_io.h
+++ b/src/core_io.h
@@ -46,7 +46,7 @@ bool DecodeHexBlockHeader(CBlockHeader&, const std::string& hex_header);
  */
 bool ParseHashStr(const std::string& strHex, uint256& result);
 std::vector<unsigned char> ParseHexUV(const UniValue& v, const std::string& strName);
-int ParseSighashString(const UniValue& sighash);
+int ParseSighashString(const std::string& sighash);
 
 // core_write.cpp
 UniValue ValueFromAmount(const CAmount amount);
diff --git a/src/core_read.cpp b/src/core_read.cpp
index 84cd559b7f..31f121206c 100644
--- a/src/core_read.cpp
+++ b/src/core_read.cpp
@@ -252,26 +252,20 @@ std::vector<unsigned char> ParseHexUV(const UniValue& v, const std::string& strN
     return ParseHex(strHex);
 }
 
-int ParseSighashString(const UniValue& sighash)
+int ParseSighashString(const std::string& sighash_str)
 {
-    int hash_type = SIGHASH_DEFAULT;
-    if (!sighash.isNull()) {
-        static std::map<std::string, int> map_sighash_values = {
-            {std::string("DEFAULT"), int(SIGHASH_DEFAULT)},
-            {std::string("ALL"), int(SIGHASH_ALL)},
-            {std::string("ALL|ANYONECANPAY"), int(SIGHASH_ALL|SIGHASH_ANYONECANPAY)},
-            {std::string("NONE"), int(SIGHASH_NONE)},
-            {std::string("NONE|ANYONECANPAY"), int(SIGHASH_NONE|SIGHASH_ANYONECANPAY)},
-            {std::string("SINGLE"), int(SIGHASH_SINGLE)},
-            {std::string("SINGLE|ANYONECANPAY"), int(SIGHASH_SINGLE|SIGHASH_ANYONECANPAY)},
-        };
-        const std::string& strHashType = sighash.get_str();
-        const auto& it = map_sighash_values.find(strHashType);
-        if (it != map_sighash_values.end()) {
-            hash_type = it->second;
-        } else {
-            throw std::runtime_error(strHashType + " is not a valid sighash parameter.");
-        }
+    static std::map<std::string, int> map_sighash_values = {
+        {std::string("DEFAULT"), int(SIGHASH_DEFAULT)},
+        {std::string("ALL"), int(SIGHASH_ALL)},
+        {std::string("ALL|ANYONECANPAY"), int(SIGHASH_ALL|SIGHASH_ANYONECANPAY)},
+        {std::string("NONE"), int(SIGHASH_NONE)},
+        {std::string("NONE|ANYONECANPAY"), int(SIGHASH_NONE|SIGHASH_ANYONECANPAY)},
+        {std::string("SINGLE"), int(SIGHASH_SINGLE)},
+        {std::string("SINGLE|ANYONECANPAY"), int(SIGHASH_SINGLE|SIGHASH_ANYONECANPAY)},
+    };
+    const auto& it{map_sighash_values.find(sighash_str)};
+    if (it != map_sighash_values.end()) {
+        return it->second;
     }
-    return hash_type;
+    throw std::runtime_error(sighash_str + " is not a valid sighash parameter.");
 }
diff --git a/src/rpc/rawtransaction.cpp b/src/rpc/rawtransaction.cpp
index eb0200ccf5..e0e30b2d9e 100644
--- a/src/rpc/rawtransaction.cpp
+++ b/src/rpc/rawtransaction.cpp
@@ -1964,7 +1964,7 @@ RPCHelpMan descriptorprocesspsbt()
         EvalDescriptorStringOrObject(descs[i], provider, /*expand_priv=*/true);
     }
 
-    int sighash_type = ParseSighashString(request.params[2]);
+    int sighash_type{request.params[2].isNull() ? SIGHASH_DEFAULT: ParseSighashString(request.params[2].get_str())};
     bool bip32derivs = request.params[3].isNull() ? true : request.params[3].get_bool();
     bool finalize = request.params[4].isNull() ? true : request.params[4].get_bool();
 
diff --git a/src/rpc/rawtransaction_util.cpp b/src/rpc/rawtransaction_util.cpp
index 3a6fa39e4d..7272cca1c5 100644
--- a/src/rpc/rawtransaction_util.cpp
+++ b/src/rpc/rawtransaction_util.cpp
@@ -289,7 +289,7 @@ void ParsePrevouts(const UniValue& prevTxsUnival, FillableSigningProvider* keyst
 
 void SignTransaction(CMutableTransaction& mtx, const SigningProvider* keystore, const std::map<COutPoint, Coin>& coins, const UniValue& hashType, UniValue& result)
 {
-    int nHashType = ParseSighashString(hashType);
+    int nHashType{hashType.isNull() ? SIGHASH_DEFAULT : ParseSighashString(hashType.get_str())};
 
     // Script verification errors
     std::map<int, bilingual_str> input_errors;
diff --git a/src/test/fuzz/parse_univalue.cpp b/src/test/fuzz/parse_univalue.cpp
index bfa856211d..5cb2c64f26 100644
--- a/src/test/fuzz/parse_univalue.cpp
+++ b/src/test/fuzz/parse_univalue.cpp
@@ -73,10 +73,6 @@ FUZZ_TARGET(parse_univalue, .init = initialize_parse_univalue)
     } catch (const UniValue&) {
     } catch (const std::runtime_error&) {
     }
-    try {
-        (void)ParseSighashString(univalue);
-    } catch (const std::runtime_error&) {
-    }
     try {
         (void)AmountFromValue(univalue);
     } catch (const UniValue&) {
diff --git a/src/wallet/rpc/spend.cpp b/src/wallet/rpc/spend.cpp
index b695d4bed3..3d74b12ef7 100644
--- a/src/wallet/rpc/spend.cpp
+++ b/src/wallet/rpc/spend.cpp
@@ -943,7 +943,7 @@ RPCHelpMan signrawtransactionwithwallet()
     // Parse the prevtxs array
     ParsePrevouts(request.params[1], nullptr, coins);
 
-    int nHashType = ParseSighashString(request.params[2]);
+    int nHashType{request.params[2].isNull() ? SIGHASH_DEFAULT: ParseSighashString(request.params[2].get_str())};
 
     // Script verification errors
     std::map<int, bilingual_str> input_errors;
@@ -1587,7 +1587,7 @@ RPCHelpMan walletprocesspsbt()
     }
 
     // Get the sighash type
-    int nHashType = ParseSighashString(request.params[2]);
+    int nHashType{request.params[2].isNull() ? SIGHASH_DEFAULT: ParseSighashString(request.params[2].get_str())};
 
     // Fill transaction with our data and also sign
     bool sign = request.params[1].isNull() ? true : request.params[1].get_bool();

```
</details>

sedited committed:
kernel: Split ParseSighashString

This split is done in preparation for the next commit where the
dependency on UniValue in the kernel library is removed.

sedited committed:
kernel: Remove Univalue from kernel library

It is not required by any of the kernel components.
A JSON library should not need to be part of a consensus library.

TheCharlatan commented:
Updated 03e06e589a24f0bcb7dbd28a0aaa925b18b87b9d -> 0498ea61f4ce1d73e84f866f420a2afb209fb01c ([kernelRmUnivalue_0](https://github.com/TheCharlatan/bitcoin/tree/kernelRmUnivalue_0) -> [kernelRmUnivalue_1](https://github.com/TheCharlatan/bitcoin/tree/kernelRmUnivalue_1),  [compare](https://github.com/TheCharlatan/bitcoin/compare/kernelRmUnivalue_0..kernelRmUnivalue_1))

* Addressed @MarcoFalke's [comment](https://github.com/bitcoin/bitcoin/pull/28113#discussion_r1269708017), fixed include
* Addressed @MarcoFalke's [comment](https://github.com/bitcoin/bitcoin/pull/28113#discussion_r1269712862), exported `univalue.h` include
* Followed up on @stickies-v's [comment](https://github.com/bitcoin/bitcoin/pull/28113#discussion_r1269745657), implemented an alternative to the comment by splitting the implementation between `univalue_helpers` and `core_read` functions. Changed the return type of the function in `core_read` to return a `util::Result` instead of throwing. Since other `UniValue` calls may throw already in the `univalue_helpers` function, it seems more appropriate to throw there though.

TheCharlatan commented:
Updated 0498ea61f4ce1d73e84f866f420a2afb209fb01c -> 05477b5566dc6f9c3d5c9609268002c0fbe1e1aa ([kernelRmUnivalue_1](https://github.com/TheCharlatan/bitcoin/tree/kernelRmUnivalue_1) -> [kernelRmUnivalue_2](https://github.com/TheCharlatan/bitcoin/tree/kernelRmUnivalue_2),  [compare](https://github.com/TheCharlatan/bitcoin/compare/kernelRmUnivalue_1..kernelRmUnivalue_2))

* Addressed @MarcoFalke's [comment](https://github.com/bitcoin/bitcoin/pull/28113#discussion_r1270790467), fixed include spacing.

fanquake commented:
Concept ACK. cc @theuni

hebasto commented:
Putting new parsing helpers into the `libbitcoin_common` library is a bit questionable for me.

I was expecting to see them in the `libbitcoin_util`. However, it is not possible with the suggested approach as the `univalue_helpers.cpp` depends on `SIGHASH_DEFAULT`.

hebasto commented:
The `ParseHexUV` function is used in `bitcoin-tx.cpp` only. Why not make it `static`/`namespace`d in there?

See: https://github.com/hebasto/bitcoin/commit/6d0fcfe3abc8b944d296d2033b9fd2e444d85ce4

TheCharlatan commented:
> The ParseHexUV function is used in bitcoin-tx.cpp only. Why not make it static/namespaced in there?

Sacrificing the fuzz test seems unfortunate. but we already don't fuzz a similar function in `bitcoin-tx` `AmountFromValue`. I guess we have poor coverage of `bitcoin-tx` in general though.

hebasto commented:
> Sacrificing the fuzz test seems unfortunate.

Maybe more fuzzing for `ParseHex` and `IsHex` is better way?

hebasto commented:
Suggesting an alternative implementation that  is based on the following idea:
```diff
--- a/src/core_read.cpp
+++ b/src/core_read.cpp
@@ -242,10 +242,10 @@ bool ParseHashStr(const std::string& strHex, uint256& result)
     return true;
 }
 
-int ParseSighashString(const UniValue& sighash)
+int ParseSighashString(const std::optional<std::string>& sighash)
 {
     int hash_type = SIGHASH_DEFAULT;
-    if (!sighash.isNull()) {
+    if (sighash.has_value()) {
         static std::map<std::string, int> map_sighash_values = {
             {std::string("DEFAULT"), int(SIGHASH_DEFAULT)},
             {std::string("ALL"), int(SIGHASH_ALL)},
@@ -255,7 +255,7 @@ int ParseSighashString(const UniValue& sighash)
             {std::string("SINGLE"), int(SIGHASH_SINGLE)},
             {std::string("SINGLE|ANYONECANPAY"), int(SIGHASH_SINGLE|SIGHASH_ANYONECANPAY)},
         };
-        const std::string& strHashType = sighash.get_str();
+        const std::string& strHashType = sighash.value();
         const auto& it = map_sighash_values.find(strHashType);
         if (it != map_sighash_values.end()) {
             hash_type = it->second;
```

The full branch is here: https://github.com/hebasto/bitcoin/tree/230721-univalue.3.

Diffs are simpler and smaller. A new helper does not depend on symbols from the `libbitcoin_consensus` library.

Feel free to take it :)

TheCharlatan commented:
Updated 05477b5566dc6f9c3d5c9609268002c0fbe1e1aa -> 67e172a7c0d82271a13f77f76ce72799faddd05c ([kernelRmUnivalue_2](https://github.com/TheCharlatan/bitcoin/tree/kernelRmUnivalue_2) -> [kernelRmUnivalue_3](https://github.com/TheCharlatan/bitcoin/tree/kernelRmUnivalue_3),  [compare](https://github.com/TheCharlatan/bitcoin/compare/kernelRmUnivalue_2..kernelRmUnivalue_3))

* Got rid of the new `univalue_helpers` file again.
* Refactored the first commit to move the UniValue-specific code to `rpc/util`.
* Moved the logic in the second commit to `bitcoin-tx.cpp`, sacrificing the fuzz test.

TheCharlatan commented:
Updated 67e172a7c0d82271a13f77f76ce72799faddd05c -> b89567f51ade926af8c918e4787046b7ccec8eb0 ([kernelRmUnivalue_3](https://github.com/TheCharlatan/bitcoin/tree/kernelRmUnivalue_3) -> [kernelRmUnivalue_4](https://github.com/TheCharlatan/bitcoin/tree/kernelRmUnivalue_4),  [compare](https://github.com/TheCharlatan/bitcoin/compare/kernelRmUnivalue_3..kernelRmUnivalue_4))

* Fixed fuzz test by catching the UniValue exception.

hebasto reviewed (commented):
Approach ACK b89567f51ade926af8c918e4787046b7ccec8eb0. Nice!

@TheCharlatan 

Do I understand you correctly that your intention is to make the `ParseSighash` quite generic?

TheCharlatan commented:
Re https://github.com/bitcoin/bitcoin/pull/28113#pullrequestreview-1542134637

> Do I understand you correctly that your intention is to make the ParseSighash quite generic?

Maybe, but I mostly thought it might be worthwhile to keep this utility around.

jonatack reviewed (commented):
Approach ACK b89567f51ade926af8c918e4787046b7ccec8eb0

TheCharlatan commented:
Updated b89567f51ade926af8c918e4787046b7ccec8eb0 -> a3774d1b2a5ce9aa6d6d3cedc2c9b9a5d2f68240 ([kernelRmUnivalue_4](https://github.com/TheCharlatan/bitcoin/tree/kernelRmUnivalue_4) -> [kernelRmUnivalue_5](https://github.com/TheCharlatan/bitcoin/tree/kernelRmUnivalue_5),  [compare](https://github.com/TheCharlatan/bitcoin/compare/kernelRmUnivalue_4..kernelRmUnivalue_5))

* Addressed @jonatack's [comment](https://github.com/bitcoin/bitcoin/pull/28113#discussion_r1271533741), fixing IWYU.
* Addressed @jonatack's [comment](https://github.com/bitcoin/bitcoin/pull/28113#discussion_r1271543977), renaming variable holding result type.
* Addressed @jonatack's [comment](https://github.com/bitcoin/bitcoin/pull/28113#discussion_r1271496876), applying suggestion on optional usage.

Did not apply the suggestions that would lead make the code moves less pure.

hebasto commented:
I've restarted the "Win64 native [vs2022]" CI task due to a remote Chocolatey issue.

stickies-v reviewed (commented):
Approach ACK a3774d1b2a5ce9aa6d6d3cedc2c9b9a5d2f68240

I think the `rpc/util` approach is much better than before. Would prefer seeing the `ParseSighash` signature simplified (and renamed), but not necessarily a blocker.

> Should throwing RPC_INVALID_PARAMETER be mentioned in the release notes?

The only behaviour change I can see here is the return code changing from -1 (RPC_MISC_ERROR) to -8 (RPC_INVALID_PARAMETER), the latter being strictly more helpful and I don't think this is a backwards incompatible change. The message and other behaviour are the same. I think a release note is not really necessary here, but obviously doesn't hurt.

TheCharlatan commented on src/core_io.h:
```diff
@@ -45,8 +47,7 @@ bool DecodeHexBlockHeader(CBlockHeader&, const std::string& hex_header);
 bool ParseHashStr(const std::string& strHex, uint256& result);
-std::vector<unsigned char> ParseHexUV(const UniValue& v, const std::string& strName);
-int ParseSighashString(const UniValue& sighash);
+util::Result<int> ParseSighash(const std::optional<std::string>& sighash);
```
The optional behavior was requested by @hebasto to isolate the sighash-specific logic from the univalue-specific logic. Could just add a docstring instead to the `ParseSighash` function?

stickies-v replied:
I think that comment was from when we still had the univalue helpers? 

Now, we have a string parsing function (in `core_read.cpp`) and an rpc helper function (in `rpc/util.cpp`). The string parsing, imo, should do just that, and the rpc helper can inject rpc-specific logic, being that 1) we like using UniValues in the RPC and that 2) for all rpcs, we prefer to interpret missing sighash as SIGHASH_DEFAULT. I don't think it's ideal to make the string parsing opinionated.

stickies-v replied:
Btw if the main concern is with `#include <script/interpreter.h>` in `rpc.util.cpp`, can avoid that by just calling `ParseSighash("DEFAULT")` too.

hebasto replied:
Using string literals scattered across the code seems fragile. In this PR branch @ a3774d1b2a5ce9aa6d6d3cedc2c9b9a5d2f68240, we already have two places:
- `sighashOptions` in `bitcoin-tx.cpp`, and
- `ParseSighash` in `core_read.cpp`

I'd avoid multiplying such cases. We'd better to refactor code (as a follow up) out into a dedicated structure/function.

theuni replied:
I like this, thanks! It just looks like a more modern version of the other functions. I also like that the weird no-value case is handled at the JSON layer instead. That's indeed cleaner than what I suggested.

theuni reviewed (commented):
Concept ACK, but IMO the new helper is quite ugly and awkward :\

I left more specific comments/suggestions.

TheCharlatan commented:
Updated a3774d1b2a5ce9aa6d6d3cedc2c9b9a5d2f68240 -> 42233bea28f6f7c464f0cd6499d675e81b3e8512 ([kernelRmUnivalue_5](https://github.com/TheCharlatan/bitcoin/tree/kernelRmUnivalue_5) -> [kernelRmUnivalue_6](https://github.com/TheCharlatan/bitcoin/tree/kernelRmUnivalue_6),  [compare](https://github.com/TheCharlatan/bitcoin/compare/kernelRmUnivalue_5..kernelRmUnivalue_6))

* Addressed @stickies-v's [comment](https://github.com/bitcoin/bitcoin/pull/28113#discussion_r1272218244), pushing the handling of the default back to the function in `rpc/util.cpp`.
* Addressed @stickies-v's [comment](https://github.com/bitcoin/bitcoin/pull/28113#discussion_r1272088375), renaming the function in `core_read.cpp` to `SighashFromStr`. There is already a function `SighashToStr` defined in `core_io.h`, so I thought having a corresponding declaration makes sense. 
* Added `[[nodiscard]]` qualifier to `SighashFromStr`.
* Addressed @stickies-v's [comment](https://github.com/bitcoin/bitcoin/pull/28113#discussion_r1272097256), but only added a shorter docstring to `ParseSighashString`.

theuni commented on src/Makefile.am:
```diff
@@ -896,8 +896,8 @@ if BUILD_BITCOIN_KERNEL_LIB
 libbitcoinkernel_la_LDFLAGS = $(AM_LDFLAGS) -no-undefined $(RELDFLAGS) $(PTHREAD_FLAGS)
-libbitcoinkernel_la_LIBADD = $(LIBBITCOIN_CRYPTO) $(LIBUNIVALUE) $(LIBLEVELDB) $(LIBMEMENV) $(LIBSECP256K1)
-libbitcoinkernel_la_CPPFLAGS = $(AM_CPPFLAGS) -I$(builddir)/obj -I$(srcdir)/secp256k1/include -DBUILD_BITCOIN_INTERNAL $(BOOST_CPPFLAGS) $(LEVELDB_CPPFLAGS) -I$(srcdir)/$(UNIVALUE_INCLUDE_DIR_INT)
+libbitcoinkernel_la_LIBADD = $(LIBBITCOIN_CRYPTO) $(LIBLEVELDB) $(LIBMEMENV) $(LIBSECP256K1)
```
Woohoo!

theuni reviewed (approved):
ACK 42233bea28f6f7c464f0cd6499d675e81b3e8512

Assuming CI is happy.

Very nice removal :)

jonatack reviewed (commented):
ACK modulo a few comments

TheCharlatan commented:
Updated 42233bea28f6f7c464f0cd6499d675e81b3e8512 -> c5fac53bca920626843723869a7677cef639df4d ([kernelRmUnivalue_6](https://github.com/TheCharlatan/bitcoin/tree/kernelRmUnivalue_6) -> [kernelRmUnivalue_7](https://github.com/TheCharlatan/bitcoin/tree/kernelRmUnivalue_7),  [compare](https://github.com/TheCharlatan/bitcoin/compare/kernelRmUnivalue_6..kernelRmUnivalue_7))

* Addressed @jonatack's [comment](https://github.com/bitcoin/bitcoin/pull/28113#discussion_r1272913872) and [comment](https://github.com/bitcoin/bitcoin/pull/28113#discussion_r1272913933), removing unused include.
* Addressed @jonatack's [comment](https://github.com/bitcoin/bitcoin/pull/28113#discussion_r1272919241), correcting docstring.

TheCharlatan commented on src/rpc/util.cpp:
```diff
@@ -3,15 +3,18 @@
+#include <core_io.h>
 #include <common/args.h>
 #include <consensus/amount.h>
+#include <script/interpreter.h>
```
I'll leave this for a follow-up.

stickies-v replied:
I don't think degrading our interface to save a bit of compilation time is an improvement tbh. I didn't look into it in detail but can see moving the sighash types into a separate header being a worthwhile follow-up, though.

stickies-v commented on src/test/fuzz/parse_univalue.cpp:
```diff
@@ -75,6 +68,7 @@ FUZZ_TARGET_INIT(parse_univalue, initialize_parse_univalue)
     try {
         (void)ParseSighashString(univalue);
+    } catch (const UniValue&) {
     } catch (const std::runtime_error&) {
```
I think this catch now needs to go?

TheCharlatan replied:
I'll test this.

TheCharlatan replied:
Seems like `std::runtime_error` is still thrown: https://github.com/TheCharlatan/bitcoin/commits/kernelRmUnivalue_8.

TheCharlatan replied:
Mmh, we don't do this for any of the other `get_str()` calls, so this might be better for a follow up? Would probably also be good to re-use the text from the type error.

stickies-v replied:
> Mmh, we don't do this for any of the other get_str() calls

I think we do? E.g. https://github.com/bitcoin/bitcoin/blob/e35fb7bc48d360585b80d0c7f89ac5087c1d405e/src/rpc/util.cpp#L89-L92 and https://github.com/bitcoin/bitcoin/blob/e35fb7bc48d360585b80d0c7f89ac5087c1d405e/src/rpc/util.cpp#L1199

TheCharlatan replied:
Ah right, I just looked at the wrong examples :P - will push a patch for this.

TheCharlatan commented:
Updated c5fac53bca920626843723869a7677cef639df4d -> d87edf309ffe60ae044a57ce9a0d9cc711edc365 ([kernelRmUnivalue_7](https://github.com/TheCharlatan/bitcoin/tree/kernelRmUnivalue_7) -> [kernelRmUnivalue_8](https://github.com/TheCharlatan/bitcoin/tree/kernelRmUnivalue_8),  [compare](https://github.com/TheCharlatan/bitcoin/compare/kernelRmUnivalue_7..kernelRmUnivalue_8))

* Addressed @jonatack's [comment](https://github.com/bitcoin/bitcoin/pull/28113#discussion_r1272918229), introduced `sighashtype` header. Can just drop it again, if other find it too controversial.
* Addressed @stickies-v's [comment](https://github.com/bitcoin/bitcoin/pull/28113#discussion_r1273298422), removed `std::runtime_error` catch in fuzz test by checking if the input is a string and throwing `RPC_INVALID_PARAMETER` if it is not. 
* Added a release note to reflect this change in error return type.

stickies-v commented on src/core_read.cpp:
```diff
@@ -242,36 +242,21 @@ bool ParseHashStr(const std::string& strHex, uint256& result)
-            throw std::runtime_error(strHashType + " is not a valid sighash parameter.");
-        }
+    static std::map<std::string, int> map_sighash_values = {
+        {std::string("DEFAULT"), int(SIGHASH_DEFAULT)},
```
nit: missing `#include <script/sighashtype.h>`?

stickies-v commented on doc/release-notes-28113.md:
```diff
@@ -0,0 +1,6 @@
+RPC Wallet
+----------
+
+- The `signrawtransactionwithkey` and `descriptorprocesspsbt` calls now return
```
nit: `signrawtransactionwithwallet` and `walletprocesspsbt` also affected

TheCharlatan replied:
Ugh, how did I miss those, will re-push.

fanquake commented:
Sorry for turning up late to the discussion, but I fail to see how adding the `<script/interpreter.h>`
include into `src/rpc/util.cpp` (the discussion in this thread: https://github.com/bitcoin/bitcoin/pull/28113#discussion_r1272918229) would make any difference to compile times, given it's contents is already contained in the preprocessed output of `rpc/libbitcoin_common_a-util.o`?

If you compare the preprocessed output of `rpc/libbitcoin_common_a-util.o` on master (e35fb7bc48d360585b80d0c7f89ac5087c1d405e) and master + `<script/interpreter.h>` in src/rpc/util.cpp, the difference in the size of the preprocessed output is a single line:
```bash
cat rpc/libbitcoin_common_a-util.o_master | wc -l
   96263
at rpc/libbitcoin_common_a-util.o_include | wc -l
   96264
```

Looking at the actual diff, that extra line in the include case is a newline:
```diff
diffoscope rpc/libbitcoin_common_a-util.o_master rpc/libbitcoin_common_a-util.o_include
--- rpc/libbitcoin_common_a-util.o_master
+++ rpc/libbitcoin_common_a-util.o_include
@@ -94954,14 +94954,15 @@
 uint256 DescriptorID(const Descriptor& desc);
 # 12 "rpc/util.cpp" 2
 
 
 
 
 
+
 # 1 "./util/translation.h" 1
 # 18 "./util/translation.h"
 struct bilingual_str {
```

and any other difference in the output is line number related, i.e:
```diff
@@ -96150,15 +96151,15 @@
         std::string res;
         for (const auto& i : m_inner) {
             res += i.ToString(oneline) + ",";
         }
         return "[" + res + "...]";
     }
     }
-    throw NonFatalCheckError( "Unreachable code reached (non-fatal)", "rpc/util.cpp", 1151, __func__);
+    throw NonFatalCheckError( "Unreachable code reached (non-fatal)", "rpc/util.cpp", 1152, __func__);
 }
 
 static std::pair<int64_t, int64_t> ParseRange(const UniValue& value)
 {
     if (value.isNum()) {
         return {0, value.getInt<int64_t>()};
     }
@@ -96238,15 +96239,15 @@
 
     return servicesNames;
 }
 
 
 [[nodiscard]] static UniValue BilingualStringsToUniValue(const std::vector<bilingual_str>& bilingual_strings)
 {
-    inline_check_non_fatal(!bilingual_strings.empty(), "rpc/util.cpp", 1239, __func__, "!bilingual_strings.empty()");
+    inline_check_non_fatal(!bilingual_strings.empty(), "rpc/util.cpp", 1240, __func__, "!bilingual_strings.empty()");
     UniValue result{UniValue::VARR};
     for (const auto& s : bilingual_strings) {
         result.push_back(s.original);
     }
     return result;
 }
```

Just for reference, the preprocessed size of the current change is `96391` (bigger than master) and it still contains `script/interpreter.h`.

Putting all of that aside, I have no issues with the current state of the PR, but in general wanted to mention this here, because I don't think we should be too concerned about adding or removing headers, unless it's very clear that doing so would actually make some significant benefit, i.e dropping Boost or other very heavy includes, clearer code separation etc, and someone has actually bothered to check.

Obviously we should be avoiding making code or interface/design decisions based on whether or not including a header might slightly change compile times for somebody trying to compile Core on a Raspberry Pi. Especially when it turns out adding or removing that header makes 0 actual difference.

stickies-v reviewed (approved):
ACK d87edf309ffe60ae044a57ce9a0d9cc711edc365

Removes dependency on UniValue in kernel, and imo improves the interface of dealing with sighashtype user input along the way.

No strong view on keeping/reverting the separate `script/sighashtype.h` header, happy to go either way. Unnecessary code churn is always better to be avoided, but I'd also rather not bike-shed over it. Will quickly re-ACK if consensus is to revert it, though.

theuni reviewed (approved):
Re-ACK 6960c81cbfa6208d4098353e53b313e13a21cb49

theuni commented:
I think the header split was reasonable btw, and I think it's safe to assume it may be redone in the future, but not for the stated reasons. Breaking something like that out for the sake of modularity and reducing dependencies is fine, but I agree with @fanquake that even if it did speed up compile, that alone would not be a compelling enough reason to reorganize the code.

stickies-v reviewed (approved):
re-ACK https://github.com/bitcoin/bitcoin/commit/6960c81cbfa6208d4098353e53b313e13a21cb49

achow101 commented:
ACK 6960c81cbfa6208d4098353e53b313e13a21cb49

TheCharlatan commented on doc/release-notes-28113.md:
```diff
@@ -0,0 +1,7 @@
+- The `signrawtransactionwithkey`, `signrawtransactionwithwallet`,
+  `walletprocesspsbt` and `descriptorprocesspsbt` calls now return more
+  specific RPC_INVALID_PARAMETER instead of RPC_PARSE_ERROR if their
+  sighashtype argument is malformed or not a string.
```
Yeah, this is very wrong :(

Before a `RPC_PARSE_ERROR` was raised if the argument is the wrong type, and a `RPC_MISC_ERROR` if the sighash type could not be parsed.

Now a `RPC_PARSE_ERROR` is still raised if the argument is the wrong type, and a `RPC_INVALID_PARAMETER` if the sighash type could not be parsed. 

Will fix.

TheCharlatan replied:
https://github.com/bitcoin/bitcoin/pull/28162

TheCharlatan commented on src/rpc/util.cpp:
```diff
@@ -310,6 +313,21 @@ UniValue DescribeAddress(const CTxDestination& dest)
+        return SIGHASH_DEFAULT;
+    }
+    if (!sighash.isStr()) {
+        throw JSONRPCError(RPC_INVALID_PARAMETER, "sighash needs to be null or string");
```
Do you have a suggestion for how to handle the fuzz test then? Could just remove this line again and re-add the `runtime_error` catch in the test.

TheCharlatan replied:
https://github.com/bitcoin/bitcoin/pull/28162

fanquake commented:
> @fanquake could you provide more info on the methodology to reproduce your data?

@jonatack I'm just running the preprocessor and comparing the to-be-compiled output. You could inject it into the compile command from `V=1`. i.e something like `/usr/bin/ccache g++ -std=c++17 -DHAVE_CONFIG_H -I. -I../src/config  -fmacro-prefix-map=/home/ubuntu/bitcoin=. -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=3 -DHAVE_BUILD_INFO -DPROVIDE_FUZZ_MAIN_FUNCTION -I. -I./minisketch/include -I./secp256k1/include -I./univalue/include -I./leveldb/include -I/usr/include -DBOOST_MULTI_INDEX_DISABLE_SERIALIZATION -DBOOST_NO_CXX98_FUNCTION_BASE   -fdebug-prefix-map=/home/ubuntu/bitcoin=. -fstack-reuse=none -Wstack-protector -fstack-protector-all -fcf-protection=full -fstack-clash-protection -Wall -Wextra -Wformat -Wformat-security -Wvla -Wredundant-decls -Wdate-time -Wduplicated-branches -Wduplicated-cond -Wlogical-op -Woverloaded-virtual -Wsuggest-override -Wimplicit-fallthrough -Wno-unused-parameter     -fno-extended-identifiers -fPIE -g -O2 -MT rpc/libbitcoin_common_a-util.o -MD -MP -MF rpc/.deps/libbitcoin_common_a-util.Tpo -c -o rpc/libbitcoin_common_a-util.o test -f 'rpc/util.cpp' || echo './' rpc/util.cpp` and add `-E` after g++.

> The implementation at https://github.com/jonatack/bitcoin/commits/2023-07-extract-sighashtype-and-scriptverify-enums, for instance, might make more of a difference (or not!) but I'm not sure of the best way to test in order to explore this better.

If the intention was to not "pull in" `script/interpreter.h`, then it'll make no difference. The relevant include chain is `key_io.h` (included in `rpc/util`) -> `script/standard.h` -> `script/interpreter.h`.

Outside of that, I'm not sure what sort of (meaningful) change we are actually trying to measure, that would have any significant effects on compile times.
